proto-lens-tests/tests/labels_test.hs

```diff
@@ -0,0 +1,59 @@
+{-# LANGUAGE OverloadedLabels #-}
+module Main where
+
+import Data.Int (Int32)
+import Prelude hiding ((.))
+import Lens.Labels ((&), (.~), (^.), (.), runLens)
+import qualified Lens.Family2
+import qualified Lens.Family
+import Proto.Canonical (Test1, Test2, Test3, Test4)
+
+import Data.Default.Class (def)
+import Data.ProtoLens (build)
+import Data.ProtoLens.TestUtil
+import Test.HUnit ((@?=))
+import Test.Framework.Providers.HUnit (testCase)
+
+main = testMain
+    [ testCase "inside" inside
+    , testCase "outside" outside
+    , testCase "fromBuild" fromBuild
+    , testCase "lensFamily" lensFamily
+    , testCase "lensFamilyCore" lensFamilyCore
+    ]
+
+-- Test that type inference works in both directions compatibly with
+-- Data.Default.Class.
+inside :: IO ()
+inside = let msg = (def :: Test3) & #c . #a .~ 42
+         in 42 @?= msg ^. #c . #a
+
+outside :: IO ()
+outside = let msg = def & #c . #a .~ 42 :: Test3
+          in 42 @?= msg ^. #c . #a
+
+fromBuild :: IO ()
+fromBuild = let msg = build (#c . #a .~ 42) :: Test3
+            in 42 @?= msg ^. #c . #a
+
+-- Test the integration with lens-family and lens-family-core.
+-- Unfortunately, "lens-family"'s setter functions such as (.~)
+-- are higher-rank and don't always do type inference well with lens-labels'
+-- HasLens instances.  Luckily 'lens-family-core' doesn't have that problem
+-- (see below), and neither do the 'lens' or 'microlens' packages.
+lensFamily :: IO ()
+lensFamily = do
+    let msg1 = def & #a .~ 42 :: Test1
+    let msg2 = def & (runLens #a :: Lens.Family2.Lens' Test1 Int32)
+                Lens.Family2..~ 42
+    let msg3 = def & runLens #a Lens.Family2..~ 42 :: Test1
+    msg1 @?= msg2
+    msg1 @?= msg3
+
+-- | lens-family-core doesn't need any extra annotations to infer the type
+-- of `msg2`.
+lensFamilyCore :: IO ()
+lensFamilyCore = let
+    msg1 = def & #a .~ 42 :: Test1
+    msg2 = def & runLens #a Lens.Family..~ 42
```

Comment:
Switched to the non-operator `set` which does probably make the test easier to parse.